### PR TITLE
Hatskier/make price aware abstract

### PR DESCRIPTION
**Pull request by hatskier**

### CODE REVIEW

File: contracts/message-based/PriceAware.sol [hunk 1]
```diff
@@ -3,49 +3,40 @@
...
 import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
-import "@openzeppelin/contracts/access/Ownable.sol";
 
-contract PriceAware is Ownable {
+abstract contract PriceAware {
   using ECDSA for bytes32;
 
-  uint256 public maxDelay = 3 * 60;
-  address private trustedSigner;
+  uint256 public maxDelay = 3 * 60; // seconds
```
Comment (jakub-wojciechowski) [inline contracts/message-based/PriceAware.sol:10]: If we create an abstract contract, let's also get rid of this variable (read a comment below).

File: contracts/message-based/PriceAware.sol [hunk 2]
```diff
@@ -3,49 +3,40 @@
...
-
-  function authorizeSigner(address _trustedSigner) external onlyOwner {
-    require(_trustedSigner != address(0));
-    trustedSigner = _trustedSigner;
+  /* ========== VIRTUAL FUNCTIONS (CAN BE OVERRIDEN) ========== */
 
-    emit TrustedSignerChanged(trustedSigner);
-  }
+  // This function must be implemented in child contract
+  function isSignerAuthorized(address _receviedSigner) internal virtual view returns (bool);
```
Comment (jakub-wojciechowski) [inline contracts/message-based/PriceAware.sol:15]: Let's create a method isTimestampValid(uint256 timestamp),
We'd be able to put additional custom logic there or utilise CONSTANTS for gas savings.

File: contracts/message-based/PriceAware.sol [hunk 3]
```diff
@@ -3,49 +3,40 @@
...
 
-  function isSignerAuthorized(address _receviedSigner) internal virtual view returns (bool) {
-    return _receviedSigner == getTrustedSigner();
+  // This function can be overriden (e.g. in Ownable contracts)
+  function setMaxDelay(uint256 _maxDelay) external virtual {
```
Comment (jakub-wojciechowski) [inline contracts/message-based/PriceAware.sol:18]: out

File: contracts/message-based/PriceAware.sol [hunk 4]
```diff
@@ -95,16 +86,18 @@ contract PriceAware is Ownable {
...
-      //Calldataload loads slots of 32 bytes
-      //The last 65 bytes are for signature + 1 for data size
-      //We load the previous 32 bytes
+      // Calldataload loads slots of 32 bytes
+      // The last 65 bytes are for signature + 1 for data size
+      // We load the previous 32 bytes
       dataTimestamp := calldataload(sub(calldatasize(), 98))
     }
 
+    // Data timestamp validation
```
Comment (jakub-wojciechowski) [inline contracts/message-based/PriceAware.sol:99]: Timestamp checking logic should be moved to a separate method

File: contracts/message-based/PriceAwareOwnable.sol [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+pragma solidity ^0.8.2;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "./PriceAware.sol";
+
+contract PriceAwareOwnable is PriceAware, Ownable {
+
+  address private trustedSigner;
+
+  function getTrustedSigner() public view virtual returns (address) {
```
Comment (jakub-wojciechowski) [inline contracts/message-based/PriceAwareOwnable.sol:12]: I'm not sure if we need this method if we handle the logic in isSignerAuthorised. 
It doesn't add much value.

**Review by jakub-wojciechowski — CHANGES_REQUESTED** [review]

File: contracts/message-based/PriceAware.sol (hunk 2 above)
Comment (hatskier) [reply contracts/message-based/PriceAware.sol:14]: I've added function `validateTimestamp` instead.
It allows to pass different error messages

`  function validateTimestamp(uint256 _receivedTimestamp) internal virtual view {
    require(
      block.timestamp > _receivedTimestamp,
      "Data with future timestamps is not allowed");
    require(
      block.timestamp - _receivedTimestamp < MAX_DELAY,
      "Data is too old");
  }`

File: contracts/message-based/PriceAware.sol (hunk 1 above)
Comment (hatskier) [reply contracts/message-based/PriceAware.sol]: Ok

File: contracts/message-based/PriceAware.sol (hunk 3 above)
Comment (hatskier) [reply contracts/message-based/PriceAware.sol]: Ok

File: contracts/message-based/PriceAwareOwnable.sol (hunk 5 above)
Comment (hatskier) [reply contracts/message-based/PriceAwareOwnable.sol]: Sure

File: contracts/message-based/PriceAware.sol (hunk 4 above)
Comment (hatskier) [reply contracts/message-based/PriceAware.sol]: Done

**Comment by jakub-wojciechowski** [comment]
I think we should have a consistent approach of validating signer and timestamp. 
I prefer the isValid... approach that return true or false, as we could pre-validate the arguments and test methods from outside. 
I see, the value in returning "data in the future" - but this is an anomaly and it's fine to revert if something that should never happen.

**Pull request merged**